sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+  malloc and free functions.
+* Memory is freed back to the source when the SYCL queue which enqueued the
+  respective allocations and frees is synchronized with, i.e. when `queue::wait`
+  or `queue::wait_and_throw` is called, unless the current total memory
+  allocated to the memory pool under the deallocation threshold, in which case
+  the memory may be retained by the pool, though this can vary depending on
+  implementation defined parameters.
+* Memory is allocated as USM memory, in one of the USM memory allocation kinds
+  enumerated in `usm::alloc`, this is specified on construction of the
+  `memory_pool` object.
+* They are associated with a specific context and one or more device(s),
+  depending on the allocation kind, this is specified on construction of the
+  `memory_pool` object.
+
+Memory pools are intended to be used for both in-order and out-of-order SYCL
+queues.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+class memory_pool {
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, Properties props = {});
+
+  template <typename Properties = empty_properties_t>
+  memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+
+  ~memory_pool();
+
+  context get_context() const;
+
+  std::vector<device> get_devices() const;
+
+  usm::alloc get_alloc_kind() const;
+
+  size_t get_max_size() const;
+
+  size_t get_threshold() const;
+
+  void set_new_threshold(size_t newThreshold);
+
+}; // memory_pool
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` and all SYCL devices
+associated with it, with the allocation kind `usm::host` and applying any
+properties in `props`.
+
+[source, c++]
+----
+template <typename Properties = empty_properties_t>
+memory_pool(context ctx, device dev, usm::alloc kind, Properties props = {});
+----
+
+_Effects_: Constructs a memory pool associated with `ctx` and `dev`, with the
+allocation kind `kind` and applying any properties in `props`.
+
+_Throws_: An exception with the `errc::invalid` error code if `kind` is
+`usm::host`.
+
+[source, c++]
+----
+~memory_pool();
+----
+
+_Effects_: If this was the last copy, signals to the SYCL runtime for the memory
+pool to be destroyed after all remaining allocations have been freed, and
+returns immediately without waiting.
+
+[source, c++]
+----
+context get_context() const;
+----
+
+_Returns_: The SYCL context associated with the memory pool.
+
+[source, c++]
+----
+std::vector<device> get_devices() const;
+----
+
+_Returns_: The SYCL device(s) associated with the memory pool.
+
+[source, c++]
+----
+usm::alloc get_alloc_kind() const;
+----
+
+_Returns_: The memory allocation kind of the memory pool.
+
+[source, c++]
+----
+size_t get_max_size() const;
+----
+
+_Returns_: The maximum size of the memory pool.
```

Comment:
What happens if the `maximum_size` property wasn't provided?  For example, what value should we return for a USM "host" memory pool?  I'm not aware that we have any "max size" value to return in this case.

I wonder if it would be better to replace this query with a more general query that just tells the properties that were provided.

Reply:
So if the `maximum_size` property isn't provided then the memory pool will adopt an implementation defined max size. For USM host allocations where there isn't really a max size I would expect this to be something like `std::numeric_limits<size_t>::max()`.

Reply:
I think there needs to be more clarity about what "max size" means.

I think the behavior would be clear if the `memory_pool` constructor actually allocated the "max size" memory.  This way, the application would have a guarantee that this amount of memory is available to the pool throughout its lifetime.  However, I don't think that's how you intent it do work.

If you don't preallocate the memory, how can you guarantee that the pool actually has "max size" memory available?  For example, what if there are two memory pools for the same device, where each specifies a "max size"?  Is there some check that the total of all the max sizes is less that the device's limit?  What if some of the device's memory is allocated outside of all pools?  This will decrease the amount of memory available to the pools for that device.  I don't see how you can guarantee that a pool will have "max size" memory available to it in these cases unless you preallocate the memnory.